### ISSUE TITLE
com.amazonaws.AmazonClientException: Unable to execute HTTP request: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.

### ISSUE THREAD

**Issue by Piedistrict** [aws/aws-sdk-android#343]
I am getting the below error when i login using cognito  AuthenticationHandler() function returns below error

javax.net.ssl.SSLHandshakeException: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.
com.amazonaws.AmazonClientException: Unable to execute HTTP request: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.

Thanks in advance

**Comment by kvasukib** [comment]
Hello @Piedistrict,

Thank you for using our SDK and reporting to us. 

What version of AWS Android SDK, Android Platform, OS and Java version are you using? 
Depending on the OS and JDK you are using, "ca-certificaes-java" package might or might not be pulled in.

Do you have a valid AWS region set in the code before you use AuthenticationHandler()? Is the region changed recently?

Thanks

**Comment by Piedistrict** [comment]
I am using com.amazonaws:aws-android-sdk-cognitoidentityprovider:2.6.2,it is happening in android 7 only in real devices,Java version 1.8.0_131.
yes region is valid ,we didn't change for long.
Before i used aws android 2.4.6 

Please suggest.Thanks in advance

**Comment by kvasukib** [comment]
Does this error occur only after upgrading to 2.6.2 version?

**Comment by Piedistrict** [comment]
No,Before also same error came.after that only i updated to latest version.

**Comment by kvasukib** [comment]
Hi @Piedistrict,

Thanks for the reply. 
Did you change or upgrade the Operating System version or the Java JDK version recently?

**Comment by Piedistrict** [comment]
Hi 
i changed the code from one machine to other machine .

No i didn't do anything like network security configuration

**Comment by Piedistrict** [comment]
My webservice call is in https,will it give problem ?and i used okhttp

**Comment by kvasukib** [comment]
Hi @Piedistrict,

Thanks for the reply. Can you give us the code snippet where you try to use AuthenticationHandler and it throws this exception?

**Comment by Piedistrict** [comment]
public class ServiceGenerator {

    private static final String BASE_URL = "https://sample.execute.amazonaws.com/dev/";

    // No need to instantiate this class.
    private ServiceGenerator() {
    }

    private static Retrofit.Builder builder =
            new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create(TwenzyApplication.getContext().getGson()));

    private static Retrofit retrofit = builder.build();

    private static HttpLoggingInterceptor logging =
            new HttpLoggingInterceptor()
                    .setLevel(HttpLoggingInterceptor.Level.BODY);

    private static OkHttpClient.Builder httpClient =
            new OkHttpClient.Builder();

    public static <S> S createService(Class<S> serviceClass, final String authToken) {
        //adding logging interceptor to okHttp client
        if (BuildConfig.ENABLE_HTTP_INTERCEPTOR) {
            if (!httpClient.interceptors().contains(logging)) {
                httpClient.addInterceptor(logging);
            }
        }

        //adding authentication interceptor to okHttp client
        if (!TextUtils.isEmpty(authToken)) {
            AuthenticationInterceptor authInterceptor = new AuthenticationInterceptor(authToken);
            if (!httpClient.interceptors().contains(authInterceptor)) {
                httpClient.addInterceptor(authInterceptor);
                builder.client(httpClient.build());
                retrofit = builder.build();
            }
        } else {
            builder.client(httpClient.build());
            retrofit = builder.build();
        }

        return retrofit.create(serviceClass);
    }
}

**Calling Class:;**
 public void login(final String userId, final String password) {
        dataManager.saveCognitoUserId(userId);
        dataManager.login(userId, new AuthenticationHandler() {
            @Override
            public void onSuccess(CognitoUserSession userSession, CognitoDevice newDevice) {
                dataManager.setAuthToken(userSession.getAccessToken().getJWTToken());
                dataManager.setIdToken(userSession.getIdToken().getJWTToken());
                processUserProfile(userSession.getIdToken().getJWTToken());
                dataManager.initializeRestClient(false);
                loginView.loginSuccess(dataManager.getIdToken());
            }

            @Override
            public void getAuthenticationDetails(AuthenticationContinuation authenticationContinuation, String userId) {
                AuthenticationDetails authenticationDetails = new AuthenticationDetails(userId, password, null);
                authenticationContinuation.setAuthenticationDetails(authenticationDetails);
                authenticationContinuation.continueTask();
            }

            @Override
            public void getMFACode(MultiFactorAuthenticationContinuation continuation) {
            }

            @Override
            public void authenticationChallenge(ChallengeContinuation continuation) {
                LogHelper.printErrorLog("Login : Auth challenge");
            }

            @Override
            public void onFailure(Exception exception) {
                loginView.hideLoading();
                if (exception instanceof UserNotConfirmedException) {
                    EventBus.getDefault().post(new EventToOTP(EventToOTP.SCREEN_LOGIN));
                } else {
                    loginView.showCredentialError(dataManager.formatException(exception));
                }
            }
        });
    }

After getAuthenticationDetails () it is going to onFailure()

**Comment by Piedistrict** [comment]
Hi 

did u got my code.i always struck in getAuthenticationDetails() .

Please help.

**Comment by Piedistrict** [comment]
i sorted out by adding the certificate access.

private static SSLSocketFactory getSSLSocketFactory() {
        try {
            // Create a trust manager that does not validate certificate chains
            final TrustManager[] trustAllCerts = new TrustManager[]{
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                            return new java.security.cert.X509Certificate[]{};
                        }
                    }
            };

            // Install the all-trusting trust manager
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
            // Create an ssl socket factory with our all-trusting manager
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            return sslSocketFactory;
        } catch (KeyManagementException | NoSuchAlgorithmException e) {
            return null;
        }

    }


--------------------------
httpClient.sslSocketFactory(getSSLSocketFactory());
        httpClient.hostnameVerifier(new HostnameVerifier() {

            @Override
            public boolean verify(String hostname, SSLSession session) {
                return true;
            }
        });
-------------------

**Comment by kassim** [comment]
should this be closed when the solution is a dangerous "trust all certificates" solution?

**Comment by minbi** [comment]
Hi @kassim ,

If you are experiencing the same or similar issue, please open a new issue. We were not able to reproduce this issue. This was most likely a network setup issue.
